presto-main/src/main/java/com/facebook/presto/operator/ScanFilterAndProjectOperator.java

```diff
@@ -290,8 +290,6 @@ private Page processPageSource()
             if (finishing) {
                 mergingOutput.finish();
             }
-
-            pageBuilderMemoryContext.setBytes(mergingOutput.getRetainedSizeInBytes());
```

Comment:
Hm, i see this line is repeated just two lines below, but maybe this was intentional that we declare memory usage before entering `getOutput()`, so that we correctly observe peak memory usage?

I trust this an OK change, but could you please comment with (or attach to commit message) some explanation?

I looked into `master` and code looks slightly different there. Equivalent of this alloc is retained but the alloc just before return is gone.

Reply:
Why this extra allocation is a problem? This line gets called per `pageSource` input page. Is it that there is a lot of small page source pages?

Reply:
Hm, it's this allocation that was a problem, not the one after `getOutput()`. I think the issue is when there are a lot of full pages: you have to allocate more memory for them, but then `getOutput()` immediately returns the page back, and the second `setBytes` call immediately deallocates the memory. The code before didn't track the memory for pages while they're in this `processPageSource` method. And since you've already added the page to the `mergingOutput`, you already would have exceeded the peak memory usage, if you were near the limit; `getOutput()` results in less or equal memory.

I'll put an explanation in the commit message, I just wanted to get this up for review quickly to see how it looked.